apps/consulting/posts/scaling-python-banking-edition.md

```diff
@@ -0,0 +1,677 @@
+for. The RAPIDS stack libraries that we explored were cuDF, CuPy, and
+blazing-SQL (though that has now been deprecated in favor of Dask SQL).
+Another tool that we tried was HEAVY.AI, formerly OmniSci. While
+HEAVY.AI was fast, it could not handle simultaneous read/write, which
+was a showstopper for us.
+
+### The "pyGPU-cuda" Algorithm
+
+GPU architecture lends itself to a pattern that many of these tools rely on.
+First, metadata is used to to determine the maximum threads per block and blocks
+per grid. Then, the data is loaded into GPU memory using this information and
+the computation is run using one of the above tools. Finally, the data is
+extracted back to the host. For more background on techniques for memory
+swapping between main RAM and GPU-RAM, see this
+[NVIDIA blog post][nvidia mem swapping post].
+
+<img
+src="/posts/scaling-python-banking-edition/pygpu-cuda-algorithm.png"
+alt="An image representing a flow diagram of the pyGPU-cuda algorith."
+/>
+
+Unfortunately, what we ended up finding was that copying the data back
+and forth between the host's memory and the GPU memory turned out to
+be rather expensive. It was expensive enough, in fact, to mitigate a lot
+of the benefits we saw in the speedup of the computation itself. On
+top of that, the increased cost of GPU-available cloud instances made
+the usage of GPUs infeasible for this project. With
+that in mind, we decided to focus on good, old distributed CPU
+architecture for our computation.
+
+## Orchestration & Aggregation
+
+With our decision made to use a distributed CPU architecture, we
+started looking at ways we could implement Dask to solve the problem. We
+choose Dask due to its native Python implementation and its
+utilization of standard Python APIs. This makes it easy to test code
+in pandas or numpy and then quickly distribute the code using Dask,
+with only slight changes. The first approach we tried was
+a simple dataframe group-by.
+
+### Group-by
+
+We thought that the dataframe approach made sense due to the ease of
+implementation: we could simply take a list of Parquet files and pass
+that list to the Dask dataframe’s `read_parquet()` function. We would then run
+the `groupby()` method on the loaded dataframe and take the sum of the resulting
+group-by object.
+
+```python
+%%time
+
+def build_graph(paths):
+    df = dd.read_parquet(paths,
+                         index=False,
+                         storage_options=storage_options)
+    gb = df.groupby(['path', 'Date']).sum()
+    return gb
+
+
+graph = build_graph(paths_5)
+fut = client.compute(graph)
+fut.result().info()
+```
+
+<img
+src="/posts/scaling-python-banking-edition/dd-groupby-dask-dashboard.png"
+alt="A picture of a jupyter notebook input and output cells, next to an image of the cell contents equivilent dask graph being run"
+width="700px"
+/>
+
+**_!!!! (We need to walk the reader through the above figure) !!!!_**
```

Comment:
Flagging this -- we need to at least briefly tour the reader through the figure.

Reply:
I added more description to the paragraph above. Let me know what you think.